### PR TITLE
Search results retrieval and cache

### PR DESCRIPTION
**Pull request by PatFawbertMills**
Needed to overhaul the search function within the app, key highlights:
- useSearch now manages responsibility of caching, the search query and any API calls
- Search is now cached in local storage - see the utility file

I'm reluctant to separate the code out further until there is need for more abstraction required.

### CODE REVIEW

File: src/api/http-common.ts [hunk 1]
```diff
@@ -31,10 +31,10 @@ class ApiClient {
       });
   }
 
-  post(url: string, values: any, config = {}) {
+  post<T>(url: string, values: any, config = {}) {
```
Comment (diversemix) [inline src/api/http-common.ts:34]: Ooo getting slick now with generics :sunglasses:

File: src/constants/cache.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+export const CACHE_NAME = "CPR_search_cache";
+export const CACHE_LIMIT = 100;
```
Comment (diversemix) [inline src/constants/cache.ts:2]: Do you have an idea of how many bytes each cache entry is? ... if its ~10k then we could make the limit more like 1000 :shrug: ... take-it-or-leave-it :smile:

Comment (PatFawbertMills) [reply src/constants/cache.ts:2]: According to this site: https://www.javainuse.com/bytesize
Each search entry is somewhere between 15 - 25kb, depending on the size of the result e.g. number of matches.

100 is a conservative limit. The maximum size of local storage is 5mb.

File: src/utils/arrayEquality.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+export const arrayOfStringdMatch = (a?: string[], b?: string[]) => {
```
Comment (diversemix) [inline src/utils/arrayEquality.ts:1]: Could you convert to a `Set` and use the set equality?

File: src/utils/daysBetween.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+export const daysBetween = function (date1: number, date2: number) {
```
Comment (diversemix) [inline src/utils/daysBetween.ts:1]: Hmm, may be time to use `moment` ... 
```
const span = moment.duration(toDate.diff(fromDate)).asDays();
```

**Review by diversemix — APPROVED** [review]
Looking great - no show stoppers just comments :smile:

File: src/constants/cache.ts (hunk 2 above)
Comment (diversemix) [reply src/constants/cache.ts:2]: Sounds good - let's not be one of those sites that take down your browser :laughing:

File: src/utils/daysBetween.ts (hunk 4 above)
Comment (PatFawbertMills) [reply src/utils/daysBetween.ts:1]: Moment has been deprecated and is not advised for new projects, I could include something like date-fns but I think for now I'll hold off from adding a date library just for this.

File: src/utils/arrayEquality.ts (hunk 3 above)
Comment (PatFawbertMills) [reply src/utils/arrayEquality.ts:1]: I like this idea, I've not used `Set` much so far will round back to it later.

**Pull request merged**